### ISSUE TITLE
mxnet.mod.BucketingModule looks like regarding default_bucket_key as batch_size

### ISSUE THREAD

**Issue by HyXieS** [dmlc/mxnet#5509]
Recently, I want to use the lstm network to do a sentiment classification task, 
I use `mxnet.rnn.FusedRNNCell` and define the `sym_gen(seq_len)` function:
```
    layout=args.layout

    def sym_gen(seq_len):
        data = mx.sym.Variable('data')
        label = mx.sym.Variable('softmax_label')
        sent_len = mx.sym.Variable('sent_len')
        embed = mx.sym.Embedding(data=data,
                                 input_dim=len(vocab),
                                 output_dim=args.num_embed,
                                 name='embed')
        outputs, _ = cell.unroll(seq_len, inputs=embed,
                                 merge_outputs=True, layout=layout)
        output = mx.sym.SequenceLast(data=outputs,
                                     sequence_length=sent_len,
                                     use_sequence_length=True,
                                     name='lstm_last_output')
        output = mx.sym.Reshape(data=output, shape=(-1, args.num_hidden*(1+args.bidirectional)))
        fc1 = mx.sym.FullyConnected(data=output, num_hidden=30, name='fc1')
        pred  = mx.sym.FullyConnected(data=fc1, num_hidden=args.nlabel, name='pred')

        sm = mx.sym.SoftmaxOutput(data=pred, label=label,
                                  normalization='batch', name='softmax')

        return (sm, ['data', 'sent_len'], ['softmax_label', ])
```
then, I create a mxnet.mod.BucketingModule instance
```
    model = mx.mod.BucketingModule(sym_gen=sym_gen,
                                   default_bucket_key=train_data_iter.default_bucket_key,
                                   context=contexts)
```

I also implement a `dataIter` similar to `mxnet.rnn.BucketSentenceIter`, 
when `layout=TN`, the `dataIter` returns
```
provide_data=[(data_name, (default_bucket_key, batch_size)), (sequence_length_name, (batch_size,))]
provide_label=[(label_name, (batch_size,))]
```
when `layout=NT`, the `dataIter` returns
```
provide_data=[(data_name, (batch_size, default_bucket_key)), (sequence_length_name, (batch_size,))]
provide_label=[(label_name, (batch_size,))]
```

However, if the `dataIter` receive a `layout=TN` and set `args.layout='TNC'` or `args.layout='NTC'`, the program cannot run, with information (batch_size = 32, default_bucket_key=50)
```
Traceback (most recent call last):
  File "lstm_model.py", line 116, in <module>
    train(args)
  File "lstm_model.py", line 103, in train
    epoch_end_callback=mx.rnn.do_rnn_checkpoint(cell, args.model_prefix, 1) if args.model_prefix else None)
  File "/usr/local/lib/python2.7/dist-packages/mxnet-0.9.4-py2.7.egg/mxnet/module/base_module.py", line 442, in fit
    for_training=True, force_rebind=force_rebind)
  File "/usr/local/lib/python2.7/dist-packages/mxnet-0.9.4-py2.7.egg/mxnet/module/bucketing_module.py", line 289, in bind
    force_rebind=False, shared_module=None, grad_req=grad_req)
  File "/usr/local/lib/python2.7/dist-packages/mxnet-0.9.4-py2.7.egg/mxnet/module/module.py", line 388, in bind
    state_names=self._state_names)
  File "/usr/local/lib/python2.7/dist-packages/mxnet-0.9.4-py2.7.egg/mxnet/module/executor_group.py", line 197, in __init__
    self.bind_exec(data_shapes, label_shapes, shared_group)
  File "/usr/local/lib/python2.7/dist-packages/mxnet-0.9.4-py2.7.egg/mxnet/module/executor_group.py", line 276, in bind_exec
    self.data_layouts = self.decide_slices(data_shapes)
  File "/usr/local/lib/python2.7/dist-packages/mxnet-0.9.4-py2.7.egg/mxnet/module/executor_group.py", line 218, in decide_slices
    + ("%s has shape %s" % (name, shape)))
AssertionError: all data must have the same batch size: batch_size = 50, but sent_len has shape (32,)
```
but `mxnet.sym.infer_shape` runs correctly

However, if the `dataIter` receive a layout `NT` and set `args.layout='NTC'` and slightly edit the `sym_gen(seq_len)` function, the program can run

It seems that the `mxnet.mod.BucketingModule` regards `default_bucket_key` as batch size

**Comment by kevinthesun** [comment]
It seems that decide_slice function doesn't get correct batch_size from dataIter:

major_axis = [DataDesc.get_batch_axis(x.layout) for x in data_shapes]

        for (name, shape), axis in zip(data_shapes, major_axis):
            if axis == -1:
                continue

            batch_size = shape[axis]

This batch_size is 50, which is generated from dataIter. There might be some issues in dataIter.

**Comment by rorschachhb** [comment]
I encountered the same problem recently.

**Comment by leezu** [comment]
This should be fixed by #6707
